test/integration/acme/orders_controller_test.go

```diff
@@ -0,0 +1,280 @@
+						APIToken: &cmmeta.SecretKeySelector{Key: testName},
+					},
+				},
+			},
+		},
+	}
+	acmeIssuer.PrivateKey.Name = testName
+	acmeIssuer.Solvers[0].DNS01.Cloudflare.APIToken.Name = testName
+	iss := gen.Issuer(testName,
+		gen.SetIssuerNamespace(testName),
+		gen.SetIssuerACME(acmeIssuer))
+
+	iss, err = cmCl.CertmanagerV1().Issuers(testName).Create(ctx, iss, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// 1. Test that a Challenge is created for a new Order.
+
+	// Create an Order CR.
+	order := gen.Order(testName,
+		gen.SetOrderIssuer(cmmeta.ObjectReference{
+			Name: testName,
+		}),
+		gen.SetOrderNamespace(testName),
+		gen.SetOrderCsr([]byte(testName)),
+		gen.SetOrderDNSNames(testName))
+
+	order, err = cmCl.AcmeV1().Orders(testName).Create(ctx, order, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Wait for the Challenge to be created.
+	var chal *cmacme.Challenge
+	err = wait.Poll(time.Millisecond*100, time.Minute, func() (done bool, err error) {
+		chals, err := cmCl.AcmeV1().Challenges(testName).List(ctx, metav1.ListOptions{})
+		l := len(chals.Items)
+		// Challenge has not been created yet
+		if l == 0 {
+			return false, nil
+		}
+		// this should never happen
+		if l > 1 {
+			return false, fmt.Errorf("expected maximum 1 challenge, got %d", l)
+		}
+		// Check that the Challenge is owned by our Order.
+		chal = &chals.Items[0]
+		if !metav1.IsControlledBy(chal, order) {
+			return false, fmt.Errorf("found an unexpected Challenge resource: %v", chal.Name)
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// 2. Test that in an edge case, where an ACME server is misbehaving and
+	// despite Challenges being valid, the ACME order status is 'pending', we
+	// re-queue the Order so that when the ACME order does become 'ready', we
+	// finalize our Order and, in the success scenario, it eventually becomes
+	// valid.
+	// https://github.com/jetstack/cert-manager/issues/2868
+
+	// Set the Challenge state to valid- the status of the ACME order remains 'pending'.
+	chal = chal.DeepCopy()
+	chal.Status.State = cmacme.Valid
+	_, err = cmCl.AcmeV1().Challenges(testName).UpdateStatus(ctx, chal, metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Sit here for a little bit checking that the Order status remains pending
+	// and also to verify that this test works.
+	var pendingOrder *cmacme.Order
+	err = wait.Poll(time.Millisecond*200, time.Second*5, func() (bool, error) {
+		pendingOrder, err = cmCl.AcmeV1().Orders(testName).Get(ctx, testName, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if pendingOrder.Status.State != cmacme.Pending {
+			return true, nil
+		}
+		return false, nil
+	})
+	switch {
+	case err == nil:
+		t.Fatalf("Expected Order to have pending status instead got: %v", pendingOrder.Status.State)
+	case err == wait.ErrWaitTimeout:
+		// 'happy case' - Order remained pending
+	default:
+		t.Fatal(err)
+	}
```

Comment:
This seems a bit fishy...I mean waiting 5 seconds for nothing to change.
What if we set some unique Reason on the Ready condition in the case where the challenges had been solved but the ACME server still reported Order pending? Then we could wait for that Ready Condition Reason here and then move on.

Reply:
> What if we set some unique Reason on the Ready condition in the case where the challenges had been solved but the ACME server still reported Order pending? Then we could wait for that Ready Condition Reason here and then move on.

Generally I think doing this would make sense.


At the moment we only set the `Reason` field on terminally failed `Order`s. I think that if we were to also set it for this intermittent state, we need to set it in all other intermittent states + when the `Order` succeeds- for consistency and to not display a wrong `Reason` after a transition from this state to i.e [this](https://github.com/jetstack/cert-manager/blob/e941307bdce1f8d731a86a00f2db14c17d6de86a/pkg/controller/acmeorders/sync.go#L179).

Could we perhaps do that + improve this test accordingly in a separate PR?
Happy to be convinced otherwise.

Also, currently `Order.status.reason` field is used in a way that is more similar to the message field on status conditions for other CRs, see i.e [here](https://github.com/jetstack/cert-manager/blob/master/pkg/controller/acmeorders/sync.go#L89). Maybe we could change this- make `Reason`s to be more 'machine readable' and add a `Order.status.message` field for more verbose messages.

Reply:
Yes, separate PR is fine by me.  Thanks for adding the TODO comment.